subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_uart.c

```diff
@@ -0,0 +1,316 @@
+		 */
+		LOG_ERR("Bad packet header detected");
+		break;
+	default:
+		LOG_ERR("Request timeout mishandled, state: %d", hc_uart->state);
+	}
+
+	res = uart_rx_disable(hc_uart->uart_dev);
+	res = uart_rx_enable(hc_uart->uart_dev, hc_uart->rx_ctx->buf,
+			     hc_uart->rx_buf_size, UART_OVERRUN_TIMEOUT_US);
+
+	hc_uart->state = UART_HOST_CMD_READY_TO_RX;
+}
+
+static void uart_callback(const struct device *dev, struct uart_event *evt, void *user_data)
+{
+	struct ec_host_cmd_uart_ctx *hc_uart = user_data;
+	size_t new_len;
+
+	switch (evt->type) {
+	case UART_RX_RDY:
+		if (hc_uart->state == UART_HOST_CMD_READY_TO_RX) {
+			hc_uart->rx_ctx->len = 0;
+			hc_uart->state = UART_HOST_CMD_RECEIVING;
+			k_work_reschedule(&hc_uart->timeout_work, UART_REQ_RX_TIMEOUT);
+		} else if (hc_uart->state == UART_HOST_CMD_PROCESSING ||
+			   hc_uart->state == UART_HOST_CMD_SENDING) {
+			LOG_ERR("UART HOST CMD ERROR: Received data while processing or sending");
+			return;
+		} else if (hc_uart->state == UART_HOST_CMD_RX_BAD ||
+			   hc_uart->state == UART_HOST_CMD_RX_OVERRUN) {
+			/* Wait for timeout if an error has been detected */
+			return;
+		}
+
+		__ASSERT(hc_uart->state == UART_HOST_CMD_RECEIVING,
+			 "UART Host Command state mishandled, state: %d", hc_uart->state);
+
+		new_len = hc_uart->rx_ctx->len + evt->data.rx.len;
+
+		if (new_len > hc_uart->rx_buf_size) {
+			/* Bad data error, set the state and wait for timeout */
+			hc_uart->state = UART_HOST_CMD_RX_BAD;
+			return;
+		}
+
+		hc_uart->rx_ctx->len = new_len;
+
+		if (hc_uart->rx_ctx->len >= sizeof(struct ec_host_cmd_request_header)) {
+			/* Buffer has request header. Check header and get data_len */
+			size_t expected_len = request_expected_size(
+				(struct ec_host_cmd_request_header *)hc_uart->rx_ctx->buf);
+
+			if (expected_len == 0 ||
+			    expected_len > hc_uart->rx_buf_size) {
+				/* Invalid expected size, set the state and wait for timeout */
+				hc_uart->state = UART_HOST_CMD_RX_BAD;
+			} else if (hc_uart->rx_ctx->len == expected_len) {
+				/* Don't wait for overrun, because it is already done
+				 * in a UART driver.
+				 */
+				(void)k_work_cancel_delayable(&hc_uart->timeout_work);
+
+				/* Disable receiving to prevent overwriting the rx buffer while
+				 * processing. Enabling receiving to a temporary buffer to detect
+				 * unexpected transfer while processing increases average handling
+				 * time ~40% so don't do that.
+				 */
+				uart_rx_disable(hc_uart->uart_dev);
+
+				/* If no data more in request, packet is complete. Start processing
+				 */
+				hc_uart->state = UART_HOST_CMD_PROCESSING;
+
+				k_sem_give(&hc_uart->rx_ctx->handler_owns);
+			} else if (hc_uart->rx_ctx->len > expected_len) {
+				/* Overrun error, set the state and wait for timeout */
+				hc_uart->state = UART_HOST_CMD_RX_OVERRUN;
+			}
+		}
+		break;
+	case UART_RX_BUF_REQUEST:
+		/* Do not provide the second buffer, because we reload DMA after every packet. */
```

Comment:
Will this cause problems if the system is heavily loaded (I have no knowledge of this system, what amounts of data would be sent, and what baud rates, etc.)?

Reply:
No, it passes the stress tests with 3Mbps. Anyway, I have no choice, without support for DMA circle mode, because of this bug https://github.com/zephyrproject-rtos/zephyr/issues/43308